### PR TITLE
[improve][io] Elastic Search Sink can now handle raw Record

### LINKED ISSUE #20619: ElasticSearch Sink Improvements
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

The current ElasticSearch sink does not work cleanly with the changes made to allow for Sink pre-processing functions, from PIP 193. Additionally, as mentioned in another [issue](https://github.com/apache/pulsar/issues/13173), the option to use the key from the record should not be tied to whether the schema is enabled or not.

These changes allow the sink to handle `Record<String>` and `Record<byte[]>` which are likely to be output from some pre-processing transform functions.  As long as the record contains a string or bytes that can be parsed as valid JSON, it should be used.  

This comes from a thread on slack with @eolivelli and @cbornet 

### Solution

modifiy method: `public Pair<String, String> extractIdAndDocument(Record<GenericObject> record) throws JsonProcessingException` to handle cases where the `Record` is `String` or `byte[]`.

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by JeffBolle**
Fixes #20619

### Motivation

Sink pre-processing functions introduced in PIP 193 (#16739) can allow for producing a Record from the function that does not have a `GenericObject`, and instead is a `String` or `byte[]`.  The current implementation of the ElasticSearch sink doesn't handle this situation cleanly.  Additionally, the setting of the id of the ElasticSearch document from the Record key is incomplete in some cases.

This stems from Slack threads with @cbornet  and @eolivelli 
### Modifications

Modified the `extractIdAndDocument` method to create a `GenericObject` from the `Record` value if it isn't already. 

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:

  - *Added unit tests in ElasticSearchRawRecordTests*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: [https://github.com/Q6Cyber/pulsar/pull/1]

### CODE REVIEW

**Comment by JeffBolle** [comment]
@michaeljmarshall @nicoloboschi @eolivelli @cbornet 
Have I missed a step in the process? Please let me know what I need to do to get the changes reviewed.  Thank you.

File: pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchRawRecordTests.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.Optional;
+
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNull;
+
+public class ElasticSearchRawRecordTests {
+    String json = "{\"c\":\"1\",\"d\":1,\"e\":{\"a\":\"a\",\"b\":true,\"d\":1.0,\"f\":1.0,\"i\":1,\"l\":10}}";
+
+    @DataProvider(name = "rawRecordSchema")
+    public Object[] stripNonPrintableCharacters() {
```
Comment (eolivelli) [inline pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchRawRecordTests.java:38]: nit: rename to "rawRecordSchema"

**Comment by lhotari** [comment]
closing and reopening to trigger a fresh test run

**Review by eolivelli — APPROVED** [review]
Great stuff!
I have left only one small comment about a method in a test, please fix it

**Review by lhotari — APPROVED** [review]
Good work @JeffBolle ! Thanks for the contribution

File: pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchRawRecordTests.java (hunk 1 above)
Comment (JeffBolle) [reply pulsar-io/elastic-search/src/test/java/org/apache/pulsar/io/elasticsearch/ElasticSearchRawRecordTests.java:38]: Got it. Sorry. Copied the test without looking too closely at that part of it. Thank you for catching this.

**Comment by JeffBolle** [comment]
> @JeffBolle Please rebase (or merge origin/master) to your PR branch and resolve the conflict.

Roger, taking care of it now.

**Comment by JeffBolle** [comment]
@Technoboy- What additional steps do I need to take to have this PR make it into a release? My organization was expecting that after it was merged to master it would be in the 3.1 release.  However it does not appear to be, though a similar PR #20741 was included.  This issue impacts both 3.1.X and 3.0.X. Thank you.

**Pull request merged**